src/main/java/org/truffleruby/core/module/ModuleNodes.java

```diff
@@ -1072,6 +1072,62 @@ protected Object constMissing(RubyModule module, String name) {
 
     }
 
+    @CoreMethod(names = "const_source_location", required = 1, optional = 1)
+    @NodeChild(value = "module", type = RubyNode.class)
+    @NodeChild(value = "name", type = RubyNode.class)
+    @NodeChild(value = "inherit", type = RubyNode.class)
+    public abstract static class ConstSourceLocationNode extends CoreMethodNode {
+
+        @Child private MakeStringNode makeStringNode = MakeStringNode.create();
+
+        @CreateCast("name")
+        protected RubyNode coerceToStringOrSymbol(RubyNode name) {
+            return ToStringOrSymbolNodeGen.create(name);
+        }
+
+        @CreateCast("inherit")
+        protected RubyNode coerceToBoolean(RubyNode inherit) {
+            return BooleanCastWithDefaultNodeGen.create(true, inherit);
+        }
+
+        @Specialization(guards = { "strings.isRubyString(name)" })
+        @TruffleBoundary
+        protected Object constSourceLocation(RubyModule module, Object name, boolean inherit,
+                @CachedLibrary(limit = "2") RubyStringLibrary strings) {
+            final ConstantLookupResult lookupResult = ModuleOperations
+                    .lookupScopedConstant(getContext(), module, strings.getJavaString(name), inherit, this, true);
+
+            return getLocation(lookupResult);
+        }
+
+        @Specialization
+        @TruffleBoundary
+        protected Object constSourceLocation(RubyModule module, RubySymbol name, boolean inherit) {
+            final ConstantLookupResult lookupResult = ModuleOperations
+                    .lookupConstantWithInherit(getContext(), module, name.getString(), inherit, this, true);
```

Comment:
Because both String and Symbol are treated the same here, and both are converted to a j.l.String, NameToJavaStringNode can be used to simplify, I'll do that.

Reply:
Ah, but they call different methods on ModuleOperations: lookupScopedConstant vs lookupConstantWithInherit.

Reply:
Yeah, per the commit message for a0e62778e2d3f73d6a039973f48fbc875d49d435, we need to treat symbols separately so that an attempt to use a symbol for scoped lookup [raises an exception](https://github.com/oracle/truffleruby/blob/addc7b7bde3fb13c17263d7632d6a1654e662ed0/spec/ruby/core/module/const_source_location_spec.rb#L190-L196).